### PR TITLE
fix(DTFS2-5912): Portal coverEndDate validation

### PR DESCRIPTION
**Pull request by Zainzzkk**
# Issue: 
* For an issued facility which has been acknowledged, an error was thrown at submission to UKEF about coverEndDates being in the past

# Changes made:
* Added check for if AIN or MIN and if facility status is acknowledged
* Added api-test
* Added e2e test

### CODE REVIEW

File: portal-api/src/v1/validation/fields/cover-end-date.js [hunk 1]
```diff
@@ -5,35 +5,49 @@ const {
   dateValidationText,
 } = require('./date');
 
-module.exports = (submittedValues, errorList) => {
-  const newErrorList = errorList;
+const { SUBMISSION_TYPE } = require('../../../constants/deal');
+const { DEAL_STATUS } = require('../../../constants/facilities');
+
+const isValidForValidation = (deal, submittedValues) => {
```
Comment (IgnatG) [inline portal-api/src/v1/validation/fields/cover-end-date.js:11]: should this read `isReadyForValidation` ?

File: e2e-tests/portal/cypress/integration/journeys/checker/cover-end-date/deal.js [hunk 2]
```diff
@@ -0,0 +1,517 @@
+const { mandatoryCriteria } = require('../../../../../../e2e-fixtures');
+const CONSTANTS = require('../../../../fixtures/constants');
+const { nowPlusDays } = require('../../../../support/utils/dateFuncs');
+
+const date = new Date();
+const datePlusDay = nowPlusDays(1);
+
+const deal = {
```
Comment (IgnatG) [inline e2e-tests/portal/cypress/integration/journeys/checker/cover-end-date/deal.js:8]: do we need this file? or can we inherit the properties from other fixture file instead?

File: e2e-tests/portal/cypress/integration/journeys/checker/cover-end-date/dealFacilityNotAcknowledged.js [hunk 3]
```diff
@@ -0,0 +1,121 @@
+const dealAllAcknowledged = require('./deal');
```
Comment (IgnatG) [inline e2e-tests/portal/cypress/integration/journeys/checker/cover-end-date/dealFacilityNotAcknowledged.js:1]: same as above, can we inherit anything else from other files?

File: portal-api/src/v1/validation/fields/cover-end-date.js (hunk 1 above)
Comment (Zainzzkk) [reply portal-api/src/v1/validation/fields/cover-end-date.js:11]: done

File: e2e-tests/portal/cypress/fixtures/deal.js [hunk 4]
```diff
@@ -0,0 +1,403 @@
...
+      },
+    ],
+    securityDetails: {
+      exporter: '',
+    },
+  },
+  ukefComments: [
+    {
+      user: {
+        username: 'DigitalService.TradeFinance@ukexportfinance.gov.uk',
```
Comment (IgnatG) [inline e2e-tests/portal/cypress/fixtures/deal.js:386]: this shouldn't be here

File: e2e-tests/portal/cypress/integration/journeys/checker/cover-end-date/cover-end-date-validation.spec.js [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+      cy.deleteDeals(ADMIN);
+      cy.insertOneDeal(dealCoverEndDatesPast, BANK1_MAKER1)
+        .then((insertedDeal) => {
+          deal = insertedDeal;
+          dealId = deal._id;
+
+          const { mockFacilities } = dealCoverEndDatesPast;
+
+          cy.createFacilities(dealId, mockFacilities, BANK1_MAKER1).then((createdFacilities) => {
+            const bonds = createdFacilities.filter((f) => f.type === 'Bond');
```
Comment (abhi-markan) [inline e2e-tests/portal/cypress/integration/journeys/checker/cover-end-date/cover-end-date-validation.spec.js:90]: Can we use constants for `Bond` and `Loan` please.

File: e2e-tests/portal/cypress/integration/journeys/checker/cover-end-date/deal.js [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+      hasBeenIssued: false,
+      ukefGuaranteeInMonths: '12',
+      guaranteeFeePayableByBank: '10.8000',
+      value: '123123.00',
+      currencySameAsSupplyContractCurrency: 'false',
+      interestMarginFee: '12',
+      coveredPercentage: '20',
+      minimumQuarterlyFee: '20',
+      ukefExposure: '24,624.60',
+      premiumFrequency: 'Monthly',
```
Comment (abhi-markan) [inline e2e-tests/portal/cypress/integration/journeys/checker/cover-end-date/deal.js:60]: Constants for `premiumFrequency` and `premiumType`.

File: e2e-tests/portal/cypress/integration/journeys/checker/cover-end-date/deal.js [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+      conversionRate: '5.5',
+      'conversionRateDate-day': '28',
+      'conversionRateDate-month': '04',
+      'conversionRateDate-year': '2022',
+      disbursementAmount: '200.00',
+      interestMarginFee: '50',
+      coveredPercentage: '60',
+      minimumQuarterlyFee: '',
+      ukefExposure: '740.40',
+      premiumType: 'At maturity',
```
Comment (abhi-markan) [inline e2e-tests/portal/cypress/integration/journeys/checker/cover-end-date/deal.js:115]: Constant for `premiumType`.

File: portal-api/src/v1/validation/fields/cover-end-date.js [hunk 8]
```diff
@@ -5,35 +5,49 @@ const {
...
-    const nowDate = moment().format('YYYY-MM-DD');
+  if (isReadyForValidation(deal, submittedValues)) {
+    if (dateHasAllValues(coverEndDateDay, coverEndDateMonth, coverEndDateYear)) {
+      const formattedDate = `${coverEndDateYear}-${coverEndDateMonth}-${coverEndDateDay}`;
+      const nowDate = moment().format('YYYY-MM-DD');
 
-    if (moment(formattedDate).isBefore(nowDate)) {
+      if (moment(formattedDate).isBefore(nowDate)) {
+        newErrorList.coverEndDate = {
+          text: 'Cover End Date must be today or in the future',
```
Comment (abhi-markan) [inline portal-api/src/v1/validation/fields/cover-end-date.js:36]: Missing fullstop at the end of the sentence please.

**Review by abhi-markan — CHANGES_REQUESTED** [review]

File: e2e-tests/portal/cypress/fixtures/constants.js [hunk 9]
```diff
@@ -42,6 +42,40 @@ const FACILITY_STAGE = {
   UNCONDITIONAL: 'Unconditional',
 };
 
+const FEE_TYPE = {
```
Comment (IgnatG) [inline e2e-tests/portal/cypress/fixtures/constants.js:45]: can we have these properties inside the `e2e-fixtures` folder?

File: e2e-tests/portal/cypress/integration/journeys/checker/cover-end-date/cover-end-date-validation.spec.js (hunk 5 above)
Comment (abhi-markan) [reply e2e-tests/portal/cypress/integration/journeys/checker/cover-end-date/cover-end-date-validation.spec.js:91]: Still pending.

**Review by abhi-markan — CHANGES_REQUESTED** [review]

File: e2e-tests/portal/cypress/integration/journeys/checker/cover-end-date/deal.js (hunk 2 above)
Comment (Zainzzkk) [reply e2e-tests/portal/cypress/integration/journeys/checker/cover-end-date/deal.js:9]: done - inherited

File: e2e-tests/portal/cypress/integration/journeys/checker/cover-end-date/dealFacilityNotAcknowledged.js (hunk 3 above)
Comment (Zainzzkk) [reply e2e-tests/portal/cypress/integration/journeys/checker/cover-end-date/dealFacilityNotAcknowledged.js]: removed

File: e2e-tests/portal/cypress/integration/journeys/checker/cover-end-date/dealFacilityNotAcknowledged.js [hunk 10]
```diff
@@ -0,0 +1,121 @@
+const mockDeal = require('../../../../fixtures/deal');
```
[the remark this answers is not in the document]
Comment (Zainzzkk) [reply e2e-tests/portal/cypress/integration/journeys/checker/cover-end-date/dealFacilityNotAcknowledged.js]: deleted file

File: e2e-tests/portal/cypress/fixtures/deal.js (hunk 4 above)
Comment (Zainzzkk) [reply e2e-tests/portal/cypress/fixtures/deal.js]: removed

File: e2e-tests/portal/cypress/integration/journeys/checker/cover-end-date/cover-end-date-validation.spec.js (hunk 5 above)
Comment (Zainzzkk) [reply e2e-tests/portal/cypress/integration/journeys/checker/cover-end-date/cover-end-date-validation.spec.js]: done

File: e2e-tests/portal/cypress/integration/journeys/checker/cover-end-date/deal.js (hunk 6 above)
Comment (Zainzzkk) [reply e2e-tests/portal/cypress/integration/journeys/checker/cover-end-date/deal.js]: done

File: e2e-tests/portal/cypress/integration/journeys/checker/cover-end-date/deal.js (hunk 7 above)
Comment (Zainzzkk) [reply e2e-tests/portal/cypress/integration/journeys/checker/cover-end-date/deal.js]: done

File: e2e-tests/portal/cypress/integration/journeys/checker/cover-end-date/dealFacilityNotAcknowledged.js [hunk 11]
```diff
@@ -0,0 +1,121 @@
...
+      },
+      conversionRate: '5.5',
+      'conversionRateDate-day': '28',
+      'conversionRateDate-month': '04',
+      'conversionRateDate-year': '2022',
+      riskMarginFee: '20',
+      coveredPercentage: '30',
+      minimumRiskMarginFee: '',
+      ukefExposure: '370.20',
+      feeType: 'At maturity',
```
[the remark this answers is not in the document]
Comment (Zainzzkk) [reply e2e-tests/portal/cypress/integration/journeys/checker/cover-end-date/dealFacilityNotAcknowledged.js]: done

**Review by abhi-markan — APPROVED** [review]

**Pull request merged**